### PR TITLE
Electra: BeaconState implementation 

### PR DESCRIPTION
**Pull request by prestonvanloon**
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

Required for Electra.

**Which issues(s) does this PR fix?**

The beacon state is blocking progress on most electra PRs. #13903  #13898 #13893 #13911 #13888

**Other notes for review**

~We might want to split this into smaller PRs. I split up the changes into logical commits that could be their on PRs, in order.~

Merge the following PRs before this can be made ready for review:
- [x] #13921
- [x] #13923
- [x] #13924
- [x] #13925

### CODE REVIEW

File: beacon-chain/core/helpers/validators.go [hunk 1]
```diff
@@ -471,3 +486,154 @@ func LastActivatedValidatorIndex(ctx context.Context, st state.ReadOnlyBeaconSta
...
+//	def is_fully_withdrawable_validator(validator: Validator, balance: Gwei, epoch: Epoch) -> bool:
+//	    """
+//	    Check if ``validator`` is fully withdrawable.
+//	    """
+//	    return (
+//	        has_execution_withdrawal_credential(validator)  # [Modified in Electra:EIP7251]
+//	        and validator.withdrawable_epoch <= epoch
+//	        and balance > 0
+//	    )
+func IsFullyWithdrawableValidator(val *ethpb.Validator, balance uint64, epoch primitives.Epoch) bool {
```
Comment (kasey) [inline beacon-chain/core/helpers/validators.go:571]: Looks like this will collide with the pr I just approved.

Comment (prestonvanloon) [reply beacon-chain/core/helpers/validators.go]: ya I rebased and this is gone now

File: beacon-chain/core/blocks/exit_test.go [hunk 2]
```diff
@@ -135,8 +135,6 @@ func TestProcessVoluntaryExits_AppliesCorrectStatus(t *testing.T) {
 }
 
 func TestVerifyExitAndSignature(t *testing.T) {
-	undo := util.HackDenebMaxuint(t)
-	defer undo()
```
Comment (prestonvanloon) [inline beacon-chain/core/blocks/exit_test.go:139]: This was renamed to HackElectraMaxuint and serves the same purpose. It's no longer needed in this particular test.

File: beacon-chain/state/interfaces.go [hunk 3]
```diff
@@ -203,6 +206,19 @@ type ReadOnlySyncCommittee interface {
 	NextSyncCommittee() (*ethpb.SyncCommittee, error)
 }
 
+type ReadOnlyElectra interface {
+	ConsolidationBalanceToConsume() (uint64, error)
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/state/interfaces.go:211]: Done in #13947, thank you!!!

File: beacon-chain/state/state-native/getters_withdrawal.go [hunk 4]
```diff
@@ -91,36 +179,14 @@ func (b *BeaconState) ExpectedWithdrawals() ([]*enginev1.Withdrawal, error) {
 			validatorIndex = 0
 		}
 	}
-	return withdrawals, nil
-}
 
-// hasETH1WithdrawalCredential returns whether the validator has an ETH1
-// Withdrawal prefix. It assumes that the caller has a lock on the state
-func hasETH1WithdrawalCredential(val *ethpb.Validator) bool {
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/getters_withdrawal.go:99]: Moved these to helpers library

File: beacon-chain/state/state-native/getters_withdrawal_test.go [hunk 5]
```diff
@@ -1,597 +1,353 @@
...
-	require.Equal(t, false, hasETH1WithdrawalCredential(v))
-	creds = []byte{params.BeaconConfig().ETH1AddressWithdrawalPrefixByte, 0xCC}
-	v = &ethpb.Validator{WithdrawalCredentials: creds}
-	require.Equal(t, true, hasETH1WithdrawalCredential(v))
-	// No Withdrawal cred
-	v = &ethpb.Validator{}
-	require.Equal(t, false, hasETH1WithdrawalCredential(v))
-}
+func TestExpectedWithdrawals(t *testing.T) {
+	for _, stateVersion := range []int{version.Capella, version.Deneb, version.Electra} {
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/getters_withdrawal_test.go:67]: I refactored TestExpectedWIthdrawals_$FORK to be a more table driven style instead of a copy and paste with the version changed.

File: beacon-chain/state/state-native/getters_withdrawal_test.go [hunk 6]
```diff
@@ -1,597 +1,353 @@
...
+					Address:        vals[3].WithdrawalCredentials[12:],
+					Amount:         params.BeaconConfig().MinDepositAmount,
+				}
+				require.DeepEqual(t, withdrawal, expected[0])
+				params.BeaconConfig().MaxValidatorsPerWithdrawalsSweep = saved
+			})
+		})
+	}
+
+	t.Run("electra all pending partial withdrawals", func(t *testing.T) {
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/getters_withdrawal_test.go:335]: This test load a ssz test from the spec tests repo as a fully hydrated example rather than hand crafting test data.

File: beacon-chain/state/state-native/hasher.go [hunk 7]
```diff
@@ -266,5 +279,52 @@ func ComputeFieldRootsWithHasher(ctx context.Context, state *BeaconState) ([][]b
 		fieldRoots[types.HistoricalSummaries.RealPosition()] = historicalSummaryRoot[:]
 	}
 
+	if state.version >= version.Electra {
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/hasher.go:282]: All of this logic is test covered by ssz_static (and other) spec tests

File: beacon-chain/state/state-native/setters_churn_test.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/prysmaticlabs/prysm/v5/testing/require"
+	"github.com/prysmaticlabs/prysm/v5/testing/util"
+)
+
+func TestExitEpochAndUpdateChurn(t *testing.T) {
+	// Load a serialized Electra state from disk.
+	// This case is copied from spectest mainnet/electra/operations/voluntary_exit/pyspec_tests/exit_existing_churn_and_churn_limit_balance/pre.ssz_snappy
+	// and that case was retrieved from the consensus-spec-tests repository at tag v1.5.0-alpha.1.
+	// The spec tests shows that the exit epoch is 262 for validator 0 performing a voluntary exit.
+	serializedBytes, err := util.BazelFileBytes("testdata/electra_exit_0.ssz_snappy")
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/setters_churn_test.go:21]: Another spec test example test data. Helped me find an off by one bug :sweat_smile:

File: beacon-chain/state/state-native/setters_payload_header.go [hunk 9]
```diff
@@ -21,6 +23,9 @@ func (b *BeaconState) SetLatestExecutionPayloadHeader(val interfaces.ExecutionDa
 
 	switch header := val.Proto().(type) {
 	case *enginev1.ExecutionPayload:
+		if b.version != version.Bellatrix {
```
Comment (prestonvanloon) [inline beacon-chain/state/state-native/setters_payload_header.go:26]: I found it odd and bad that we could have a beacon state that accepted the wrong type of execution payload, so I added that verification for each case and the appropriate test.

File: beacon-chain/state/stateutil/slice_root.go [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/prysmaticlabs/prysm/v5/encoding/ssz"
+)
+
+// SliceRoot computes the root of a slice of hashable objects.
+func SliceRoot[T ssz.Hashable](slice []T, limit uint64) ([32]byte, error) {
```
Comment (prestonvanloon) [inline beacon-chain/state/stateutil/slice_root.go:13]: This is the same logic exactly from beacon-chain/state/stateutil/historical_summaries_root.go, but parameterized / generic version

File: beacon-chain/state/stateutil/trie_helpers.go [hunk 11]
```diff
@@ -250,11 +250,11 @@ func AddInMixin(root [32]byte, length uint64) ([32]byte, error) {
 
 // Merkleize 32-byte leaves into a Merkle trie for its adequate depth, returning
 // the resulting layers of the trie based on the appropriate depth. This function
-// pads the leaves to a length of 32.
+// pads the leaves to a length of a multiple of 32.
 func Merkleize(leaves [][]byte) [][][]byte {
 	hashFunc := hash.CustomSHA256Hasher()
 	layers := make([][][]byte, ssz.Depth(uint64(len(leaves)))+1)
-	for len(leaves) != 32 {
+	for len(leaves)%32 != 0 {
```
Comment (prestonvanloon) [inline beacon-chain/state/stateutil/trie_helpers.go:257]: This was a bug where we would allocate infinite memory if the beacon state had more than 32 fields. Wow!

File: testing/util/electra.go [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+	"github.com/prysmaticlabs/prysm/v5/testing/require"
+)
+
+// HackElectraMaxuint is helpful for tests that need to set up cases where the electra fork has passed.
+// We have unit tests that assert our config matches the upstream config, where the next fork is always
+// set to MaxUint64 until the fork epoch is formally set. This creates an issue for tests that want to
+// work with slots that are defined to be after electra because converting the max epoch to a slot leads
+// to multiplication overflow.
+// Monkey patching tests with this function is the simplest workaround in these cases.
+func HackElectraMaxuint(t *testing.T) func() {
```
Comment (prestonvanloon) [inline testing/util/electra.go:17]: Moved from testing/util/deneb.go

**Review by prestonvanloon — COMMENTED** [review]
Left some commentary from self review. This PR is ready to review and is very large so it will need more than 1 approval.

File: beacon-chain/state/interfaces.go [hunk 13]
```diff
@@ -283,3 +300,19 @@ type WriteOnlySyncCommittee interface {
 	SetCurrentSyncCommittee(val *ethpb.SyncCommittee) error
 	SetNextSyncCommittee(val *ethpb.SyncCommittee) error
 }
+
+type WriteOnlyPendingBalanceDeposits interface {
+	AppendPendingBalanceDeposit(index primitives.ValidatorIndex, amount uint64) error
```
Comment (james-prysm) [inline beacon-chain/state/interfaces.go:305]: dumb question here why is this in its own interface instead of writeOnlyElectra?

File: beacon-chain/state/state-native/BUILD.bazel [hunk 14]
```diff
@@ -38,13 +47,11 @@ go_library(
         "state_trie.go",
         "types.go",
         "validator_index_cache.go",
-    ] + select({
```
Comment (james-prysm) [inline beacon-chain/state/state-native/BUILD.bazel:41]: qq why was this moved?

File: beacon-chain/state/state-native/getters_validator.go [hunk 15]
```diff
@@ -413,3 +414,54 @@ func (b *BeaconState) inactivityScoresVal() []uint64 {
...
+//	        withdrawal.amount for withdrawal in state.pending_partial_withdrawals if withdrawal.index == validator_index)
+func (b *BeaconState) PendingBalanceToWithdraw(idx primitives.ValidatorIndex) (uint64, error) {
+	if b.version < version.Electra {
+		return 0, errNotSupported("PendingBalanceToWithdraw", b.version)
+	}
+
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	// TODO: Consider maintaining this value in the state, if it's a potential bottleneck.
```
Comment (james-prysm) [inline beacon-chain/state/state-native/getters_validator.go:456]: assuming this won't be forgotten lol, hopefully can be discussed at the interop

File: beacon-chain/state/state-native/BUILD.bazel (hunk 14 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/BUILD.bazel:41]: Because it was causing gazelle to throw a warning / error

File: beacon-chain/state/state-native/getters_validator.go (hunk 15 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/getters_validator.go:456]: I am going to test it at some point. I implemented it per spec, but the way it is used causes it to be called many times. Maintaining the value on write might be an optimization but I haven't had time to bench it yet

File: beacon-chain/state/interfaces.go (hunk 13 above)
Comment (prestonvanloon) [reply beacon-chain/state/interfaces.go:305]: I think I had split something out for maxEB, but I don't remember. There is no reason, I guess

File: beacon-chain/blockchain/execution_engine.go [hunk 16]
```diff
@@ -324,8 +324,8 @@ func (s *Service) getPayloadAttribute(ctx context.Context, st state.BeaconState,
 
 	var attr payloadattribute.Attributer
 	switch st.Version() {
-	case version.Deneb:
-		withdrawals, err := st.ExpectedWithdrawals()
+	case version.Deneb, version.Electra:
```
Comment (terencechain) [inline beacon-chain/blockchain/execution_engine.go:327]: Why do we want to add `version.Electra` here? I feel like this PR should be isolated to only `Electra: BeaconState`. If we do want to add it, I'd suggest we: 
- Add a unit test to `getPayloadAttribute`
- Also add `version.Electra` to L345 below

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go [hunk 17]
```diff
@@ -127,8 +127,8 @@ func (vs *Server) getLocalPayload(ctx context.Context, blk interfaces.ReadOnlyBe
 	}
 	var attr payloadattribute.Attributer
 	switch st.Version() {
-	case version.Deneb:
-		withdrawals, err := st.ExpectedWithdrawals()
+	case version.Deneb, version.Electra:
```
Comment (terencechain) [inline beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go:130]: Same comment as above: in some places, we add the Electra version, but in some places, we do not.

File: beacon-chain/state/state-native/beacon_state_mainnet.go [hunk 18]
```diff
@@ -103,8 +115,19 @@ type beaconStateMarshalable struct {
 	NextSyncCommittee                   *ethpb.SyncCommittee                    `json:"next_sync_committee" yaml:"next_sync_committee"`
 	LatestExecutionPayloadHeader        *enginev1.ExecutionPayloadHeader        `json:"latest_execution_payload_header" yaml:"latest_execution_payload_header"`
 	LatestExecutionPayloadHeaderCapella *enginev1.ExecutionPayloadHeaderCapella `json:"latest_execution_payload_header_capella" yaml:"latest_execution_payload_header_capella"`
+	LatestExecutionPayloadHeaderDeneb   *enginev1.ExecutionPayloadHeaderDeneb   `json:"latest_execution_payload_header_deneb" yaml:"latest_execution_payload_header_deneb"`
```
Comment (terencechain) [inline beacon-chain/state/state-native/beacon_state_mainnet.go:118]: I wonder what is breaking today we don't have deneb here

File: beacon-chain/state/stateutil/trie_helpers.go (hunk 11 above)
Comment (james-prysm) [reply beacon-chain/state/stateutil/trie_helpers.go:257]: bug of the week

File: testing/util/electra.go (hunk 12 above)
Comment (james-prysm) [reply testing/util/electra.go:17]: do we need to do this each fork?

File: beacon-chain/state/state-native/getters_validator.go [hunk 19]
```diff
@@ -413,3 +414,54 @@ func (b *BeaconState) inactivityScoresVal() []uint64 {
...
+
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+
+	if i >= primitives.ValidatorIndex(len(b.validators)) {
+		return 0, errors.Wrapf(consensus_types.ErrOutOfBounds, "validator index %d does not exist", i)
+	}
+	v := b.validators[i]
+
+	return min(b.balances[i], helpers.ValidatorMaxEffectiveBalance(v)), nil
```
Comment (terencechain) [inline beacon-chain/state/state-native/getters_validator.go:438]: Should we check `b.balances` length? 
```
if i >= primitives.ValidatorIndex(len(b. balances)) {
		return 0, errors.Wrapf(consensus_types.ErrOutOfBounds, "validator index %d does not exist", i)
	}
```

File: beacon-chain/state/state-native/setters_balance_deposits.go [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+// only applies in electra or later.
+func (b *BeaconState) AppendPendingBalanceDeposit(index primitives.ValidatorIndex, amount uint64) error {
+	if b.version < version.Electra {
+		return errNotSupported("AppendPendingBalanceDeposit", b.version)
+	}
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.sharedFieldReferences[types.PendingBalanceDeposits].MinusRef()
+	b.sharedFieldReferences[types.PendingBalanceDeposits] = stateutil.NewRef(1)
```
Comment (terencechain) [inline beacon-chain/state/state-native/setters_balance_deposits.go:22]: This looks odd. Why minus then just set it to one?

File: beacon-chain/state/state-native/setters_consolidation.go [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+// Lock on the state and only applies in electra or later.
+func (b *BeaconState) AppendPendingConsolidation(val *ethpb.PendingConsolidation) error {
+	if b.version < version.Electra {
+		return errNotSupported("AppendPendingConsolidation", b.version)
+	}
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	b.sharedFieldReferences[types.PendingConsolidations].MinusRef()
+	b.sharedFieldReferences[types.PendingConsolidations] = stateutil.NewRef(1)
```
Comment (terencechain) [inline beacon-chain/state/state-native/setters_consolidation.go:22]: Same comment here. Perhaps I'm not understanding these two lines

File: beacon-chain/state/state-native/state_trie.go [hunk 22]
```diff
@@ -683,6 +703,131 @@ func InitializeFromProtoUnsafeDeneb(st *ethpb.BeaconStateDeneb) (state.BeaconSta
...
+	b.sharedFieldReferences[types.HistoricalRoots] = stateutil.NewRef(1)
+	b.sharedFieldReferences[types.Eth1DataVotes] = stateutil.NewRef(1)
+	b.sharedFieldReferences[types.Slashings] = stateutil.NewRef(1)
+	b.sharedFieldReferences[types.PreviousEpochParticipationBits] = stateutil.NewRef(1)
+	b.sharedFieldReferences[types.CurrentEpochParticipationBits] = stateutil.NewRef(1)
+	b.sharedFieldReferences[types.LatestExecutionPayloadHeaderElectra] = stateutil.NewRef(1) // New in Electra.
+	b.sharedFieldReferences[types.HistoricalSummaries] = stateutil.NewRef(1)                 // New in Capella.
+	b.sharedFieldReferences[types.PendingBalanceDeposits] = stateutil.NewRef(1)              // New in Electra.
+	b.sharedFieldReferences[types.PendingPartialWithdrawals] = stateutil.NewRef(1)           // New in Electra.
+	b.sharedFieldReferences[types.PendingConsolidations] = stateutil.NewRef(1)               // New in Electra.
```
Comment (terencechain) [inline beacon-chain/state/state-native/state_trie.go:815]: Do we need shared references for these fields? I wonder how we are considering these

**Review by terencechain — COMMENTED** [review]
`beacon-chain/state/state-native/testdata/electra_random_0.ssz_snappy` is 3MB. I wonder if we should add it. Although the cost will be permanent, it might not be bad, but we'll also test this as part of the spec test.

File: testing/util/electra.go (hunk 12 above)
Comment (prestonvanloon) [reply testing/util/electra.go:17]: For each unreleased fork, yeah. Otherwise you get an overflow when trying to determine the start slot for an fork that is scheduled for FAR_FUTURE_EPOCH

File: beacon-chain/state/state-native/setters_balance_deposits.go (hunk 20 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/setters_balance_deposits.go:22]: That's how the shared state fields work with the garbage collector.

When you want to mutate a field that may be shared across copied states:
- Subtract your reference to the shared field
- Initialize a new field reference with count 1. This will be incremented when the state is copied in the future.
- When a field ref count reaches zero, then the garbage collector / finalizer allows the field to be destroyed

File: beacon-chain/state/state-native/state_trie.go (hunk 22 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/state_trie.go:815]: These are slices of which "large" max lengths. It could be several million entries per slice so it makes sense to share these and copy-on-write instead of copying everything in beaconState.Copy()

File: beacon-chain/state/state-native/getters_validator.go (hunk 19 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/getters_validator.go:438]: Yes

File: beacon-chain/state/state-native/beacon_state_mainnet.go (hunk 18 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/beacon_state_mainnet.go:118]: hmmm yeah good point, i dont know lol

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go (hunk 17 above)
Comment (prestonvanloon) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go:130]: This PR doesn't address all of the places that need to handle electra cases in switch statement. In fact, this change is probably out of scope of this PR. I only added it since I was here due to API changes on `st.ExpectedWithdrawals()`

Comment (terencechain) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go:130]: Agree. I think we should either do it all or not do it at all. Otherwise, the disparity might be confusing and lead to missing test coverage down the line.

File: beacon-chain/state/state-native/setters_balance_deposits.go (hunk 20 above)
Comment (terencechain) [reply beacon-chain/state/state-native/setters_balance_deposits.go:22]: Thanks!

File: beacon-chain/state/state-native/getters_validator.go (hunk 19 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/getters_validator.go:440]: Done in d54497a00544a572d979f0c1123ba858135d5ab6

File: beacon-chain/state/state-native/setters_consolidation.go (hunk 21 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/setters_consolidation.go:22]: Addressed in your other comment

File: beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go (hunk 17 above)
Comment (prestonvanloon) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go:130]: OK, I'll revert it here

Comment (prestonvanloon) [reply beacon-chain/rpc/prysm/v1alpha1/validator/proposer_execution_payload.go]: Reverted in d8ab9dff976bd93297629dc21c55d3d27af13de8

File: beacon-chain/state/state-native/beacon_state_mainnet.go (hunk 18 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/beacon_state_mainnet.go:118]: Oh, this is to json-ify the beacon state, which I am guessing we never / rarely do because no one noticed this field was missing in the JSON version of the mainnet beacon state in deneb

File: beacon-chain/blockchain/execution_engine.go (hunk 16 above)
Comment (prestonvanloon) [reply beacon-chain/blockchain/execution_engine.go]: Reverted this and the other one in 2d6b28c6497cdd4e0781e41760bf45b625743e45

**Comment by prestonvanloon** [comment]
@terencechain since I copied beacon-chain/state/state-native/testdata/electra_random_0.ssz_snappy and the other test case from spectests, i suppose I could just ingest that test case from the spectest data instead of committing it to the repo. I'll add that now

**Comment by prestonvanloon** [comment]
@terencechain removed ssz_snappy files in c85b9d4db02635b4806fc2747c8f5f73bb994c74

**Review by terencechain — DISMISSED** [review]
Looks good to me. I've carefully reviewed everything except for:
1. Comparing the methods in setters line by line with the spec.
2. Understanding how the getters and setters will be used by future callers and whether the API satisfies the callers needs.

**Review by james-prysm — APPROVED** [review]
I think i'm good with this, my comments were answered as well and I tried to go line by line in the spec(hopefully I did not miss something)

File: beacon-chain/state/state-native/state_trie.go [hunk 23]
```diff
@@ -93,17 +93,33 @@ var denebFields = append(
 	types.HistoricalSummaries,
 )
 
+var electraFields = append(
+	denebFields,
```
Comment (rkapka) [inline beacon-chain/state/state-native/state_trie.go:97]: This is wrong because `denebFields` contains `LatestExecutionPayloadHeaderDeneb`

File: beacon-chain/state/state-native/state_trie.go [hunk 24]
```diff
@@ -93,17 +93,33 @@ var denebFields = append(
...
 	bellatrixSharedFieldRefCount                  = 12
 	capellaSharedFieldRefCount                    = 14
 	denebSharedFieldRefCount                      = 14
+	electraSharedFieldRefCount                    = 17
 	experimentalStatePhase0SharedFieldRefCount    = 5
 	experimentalStateAltairSharedFieldRefCount    = 5
 	experimentalStateBellatrixSharedFieldRefCount = 6
 	experimentalStateCapellaSharedFieldRefCount   = 8
 	experimentalStateDenebSharedFieldRefCount     = 8
+	experimentalStateElectraSharedFieldRefCount   = 8
```
Comment (rkapka) [inline beacon-chain/state/state-native/state_trie.go:122]: This is wrong, the value should be `11`. The name of the const is confusing, but it means "what is the number of `SharedFieldRefCount` when you enable the experimental state? The answer is the number without the experimental state minus multi-slice fields.

File: beacon-chain/state/state-native/state_trie.go [hunk 25]
```diff
@@ -891,6 +1052,10 @@ func (b *BeaconState) initializeMerkleLayers(ctx context.Context) error {
 		b.dirtyFields = make(map[types.FieldIndex]bool, params.BeaconConfig().BeaconStateCapellaFieldCount)
 	case version.Deneb:
 		b.dirtyFields = make(map[types.FieldIndex]bool, params.BeaconConfig().BeaconStateDenebFieldCount)
+	case version.Electra:
+		b.dirtyFields = make(map[types.FieldIndex]bool, params.BeaconConfig().BeaconStateElectraFieldCount)
+	default:
+		return errors.New("unknown state version when computing dirty fields in merklization")
```
Comment (rkapka) [inline beacon-chain/state/state-native/state_trie.go:1058]: can you include the string representation of `b.version` in the error message?

File: beacon-chain/state/state-native/types/types.go [hunk 26]
```diff
@@ -83,15 +85,37 @@ func (f FieldIndex) String() string {
...
+		return "earliestConsolidationEpoch"
+	case PendingBalanceDeposits:
+		return "pendingBalanceDeposits"
+	case PendingPartialWithdrawals:
+		return "pendingPartialWithdrawals"
+	case PendingConsolidations:
+		return "pendingConsolidations"
 	default:
-		return ""
+		return fmt.Sprintf("unknown field index name: %d", f)
```
Comment (rkapka) [inline beacon-chain/state/state-native/types/types.go:118]:
```suggestion
		return fmt.Sprintf("unknown field index number: %d", f)
```

File: beacon-chain/state/state-native/state_trie.go (hunk 24 above)
Comment (james-prysm) [reply beacon-chain/state/state-native/state_trie.go:122]: can that be added in some comment in the code? i didn't understand the number changes either.

File: beacon-chain/state/state-native/getters_withdrawal.go [hunk 27]
```diff
@@ -55,31 +112,62 @@ func (b *BeaconState) ExpectedWithdrawals() ([]*enginev1.Withdrawal, error) {
 	withdrawalIndex := b.nextWithdrawalIndex
 	epoch := slots.ToEpoch(b.slot)
 
+	// Electra partial withdrawals functionality.
+	if b.version >= version.Electra {
+		for _, w := range b.pendingPartialWithdrawals {
+			if w.WithdrawableEpoch > epoch || len(withdrawals) >= int(params.BeaconConfig().MaxPendingPartialsPerWithdrawalSweep) {
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/state/state-native/getters_withdrawal.go:118]: What do you mean?

File: beacon-chain/state/state-native/BUILD.bazel [hunk 28]
```diff
@@ -3,11 +3,16 @@ load("@prysm//tools/go:def.bzl", "go_library", "go_test")
 go_library(
     name = "go_default_library",
     srcs = [
+        "beacon_state_mainnet.go",
+        "beacon_state_minimal.go",  # keep
         "doc.go",
         "error.go",
         "getters_attestation.go",
+        "getters_balance_deposits.go",
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/state/state-native/BUILD.bazel:11]: I am going to leave it as is for now. Thanks

File: beacon-chain/state/state-native/setters_churn_test.go [hunk 29]
```diff
@@ -0,0 +1,38 @@
...
+
+	"github.com/golang/snappy"
+	state_native "github.com/prysmaticlabs/prysm/v5/beacon-chain/state/state-native"
+	"github.com/prysmaticlabs/prysm/v5/consensus-types/primitives"
+	eth "github.com/prysmaticlabs/prysm/v5/proto/prysm/v1alpha1"
+	"github.com/prysmaticlabs/prysm/v5/testing/require"
+	"github.com/prysmaticlabs/prysm/v5/testing/util"
+)
+
+func TestExitEpochAndUpdateChurn(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply beacon-chain/state/state-native/setters_churn_test.go:15]: I must admit, I was tired/lazy by the time i got to testing this one and I am pleased that you called me out :rofl:

**Comment by prestonvanloon** [comment]
All of @rkapka feedback is addressed. This PR is ready for another round of reviews

**Comment by prestonvanloon** [comment]
Rebased onto develop and broken by https://github.com/prysmaticlabs/prysm/pull/13948. Trying to figure how how to fix most effectively.

**Comment by prestonvanloon** [comment]
Build is fixed as of 4ff0357

**Review by rkapka — APPROVED** [review]
Fantastic work!

**Review by james-prysm — APPROVED** [review]

File: beacon-chain/state/state-native/getters_consolidation.go [hunk 30]
```diff
@@ -0,0 +1,64 @@
...
+// NumPendingConsolidations is a non-mutating call to the beacon state which returns the number of
+// pending consolidations in the beacon state. This method requires access to the RLock on the state
+// and only applies in electra or later.
+func (b *BeaconState) NumPendingConsolidations() (uint64, error) {
+	if b.version < version.Electra {
+		return 0, errNotSupported("NumPendingConsolidations", b.version)
+	}
+	b.lock.RLock()
+	defer b.lock.RUnlock()
+	return uint64(len(b.pendingConsolidations)), nil
```
Comment (kasey) [inline beacon-chain/state/state-native/getters_consolidation.go:55]: `pendingConsolidationsVal` copies the values and then they are discarded. It would make more sense to allow this method to directly check `len(b.pendingConsolidations)`.

File: beacon-chain/state/state-native/getters_withdrawal_test.go [hunk 31]
```diff
@@ -1,597 +1,352 @@
...
-			val := &ethpb.Validator{
-				WithdrawalCredentials: make([]byte, 32),
-				EffectiveBalance:      params.BeaconConfig().MaxEffectiveBalance,
-				WithdrawableEpoch:     primitives.Epoch(0),
-			}
-			val.WithdrawalCredentials[0] = params.BeaconConfig().ETH1AddressWithdrawalPrefixByte
-			s.validators[i] = val
-		}
-		expected, err := s.ExpectedWithdrawals()
+		serializedSSZ, err := snappy.Decode(nil /* dst */, serializedBytes)
```
Comment (kasey) [inline beacon-chain/state/state-native/getters_withdrawal_test.go:340]: This could be a nice general test helper (get example state for fork).

File: beacon-chain/state/state-native/getters_consolidation.go (hunk 30 above)
Comment (kasey) [reply beacon-chain/state/state-native/getters_consolidation.go:55]: Nevermind, I can't read.

File: beacon-chain/state/state-native/getters_withdrawal_test.go (hunk 31 above)
Comment (prestonvanloon) [reply beacon-chain/state/state-native/getters_withdrawal_test.go:340]: If someone finds that useful, I would be happy to help move this to some shared method. I thought this was a rather specific example that I crawled the spectests repo to find.

**Review by kasey — APPROVED** [review]
